### DISCUSSION TITLE
User has no config

### DISCUSSION THREAD

**Discussion by someconservator** [netbox-community/netbox#8718] [Q&A]
### NetBox version

v.3.1.8

### Python version

3.9

### Steps to Reproduce

1. Click "log in"
2. Enter credentials
3. Click "sign in"

### Expected Behavior

Correct login

### Observed Behavior

<class 'django.contrib.auth.models.User.config.RelatedObjectDoesNotExist'>

User has no config.

Python version: 3.9.7
NetBox version: 3.1.8

**Comment by jeremystretch** [comment]
Thank you for opening a bug report. Unfortunately, the information you have provided is not sufficient for someone else to attempt to reproduce the reported behavior. Remember, each bug report must include detailed steps that someone else can follow on a clean, empty NetBox installation to reproduce the exact problem you're experiencing. These instructions should include the creation of any involved objects, any configuration changes, and complete accounting of the actions being taken. Also be sure that your report does not reference data on the public NetBox demo, as that is subject to change at any time by an outside party and cannot be relied upon for bug reports.

**Comment by someconservator** [comment]
Excuse me, but what other steps need to be specified? This is exactly what happens to me: I try to log in and I get a window with this error.

**Comment by jeremystretch** [comment]
Per the bug report template:

> Describe in detail the exact steps that someone else can take to reproduce this bug using the current stable release of NetBox.

You have not provided any such detail. Obviously merely logging into NetBox does not reproduce the problem, as you can verify for yourself [here](https://demo.netbox.dev/). If you'd like someone to look into this, you'll need to spend some time determining how to reproduce it.

**Comment by someconservator** [comment]
During today I've narrowed down the possible issues: a fresh installation of netbox (by the way, this is centos8.5, and nginx1.20, and with ldap) does indeed work fine. My netbox installation is an upgraded installation from version 3.1. As soon as i dump its database to a fresh installation i immediately get the same error. I have a suspicion that the update script did not make some necessary migrations (while the data itself is present in the database, I checked)

**Comment by someconservator** [comment]
During today I've narrowed down the possible issues: a fresh installation of netbox (by the way, this is centos8.5, and nginx1.20, and with ldap) does indeed work fine. My netbox installation is an upgraded installation from version 3.1. As soon as i dump its database to a fresh installation i immediately get the same error. I have a suspicion that the update script did not make some necessary migrations (while the data itself is present in the database, I checked)

**Comment by someconservator** [comment]
When i try to login in DEBUG mode in test server, i get on console:
[22/Feb/2022 08:22:22] "POST /login/ HTTP/1.1" 302 0
Internal Server Error: /
Traceback (most recent call last):
  File "/opt/netbox-3.1.8/venv/lib64/python3.9/site-packages/django/core/handlers/exception.py", line 47, in inner
    response = get_response(request)
  File "/opt/netbox-3.1.8/venv/lib64/python3.9/site-packages/django/core/handlers/base.py", line 181, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/opt/netbox-3.1.8/venv/lib64/python3.9/site-packages/django/views/generic/base.py", line 70, in view
    return self.dispatch(request, *args, **kwargs)
  File "/opt/netbox-3.1.8/venv/lib64/python3.9/site-packages/django/views/generic/base.py", line 98, in dispatch
    return handler(request, *args, **kwargs)
  File "/opt/netbox-3.1.8/netbox/netbox/views/__init__.py", line 136, in get
    changelog_table = ObjectChangeTable(changelog, user=request.user)
  File "/opt/netbox-3.1.8/netbox/utilities/tables.py", line 68, in __init__
    selected_columns = user.config.get(f"tables.{self.__class__.__name__}.columns")
  File "/opt/netbox-3.1.8/venv/lib64/python3.9/site-packages/django/utils/functional.py", line 247, in inner
    return func(self._wrapped, *args)
  File "/opt/netbox-3.1.8/venv/lib64/python3.9/site-packages/django/db/models/fields/related_descriptors.py", line 421, in __get__
    raise self.RelatedObjectDoesNotExist(
django.contrib.auth.models.User.config.RelatedObjectDoesNotExist: User has no config.

How I can further understand, what gives me such an error?

**Comment by someconservator** [comment]
I found!
When migrating from an older version, the users_userconfig table remains empty: 
netboxdb=> select * from users_userconfig;
 id | data | user_id
----+------+---------
(0 rows)

when i manually do:
netboxdb=> INSERT INTO users_userconfig values (1,'{}',1);

this error disappears.

hence the questions:
why does this happen?
how best to fix this?
will this continue to happen when a new user is added?

**Reply by Jacq** [reply]
In my case it was a fresh install of Netbox version 0b508e83bd3 (v3.2.2), exactly the same issue, we could complete the first login of the admin user and we had to create manually the users_userconfig record as @someconservator.

**Reply by KyleKaniecki** [reply]
This happens with me as well

```
<class 'django.contrib.auth.models.User.config.RelatedObjectDoesNotExist'>

User has no config.

Python version: 3.9.5
NetBox version: 3.2.3
```

Logging in with `admin/admin` on a fresh install

**Reply by Guruleenyc** [reply]
Same issue as well.

**Comment by prnzngr** [comment]
I have the same issue. Any news?

**Comment by Mikesco3** [comment]
same issue here...

**Comment by cpmills1975** [comment]
I've just hit the same issue. In my case, I'm trying to build a NetBox installation running on Kubernetes on a cluster of Raspberry Pi's in my home lab. So, under the hood I am running the following containers:
* docker.io/tgenannt/netbox:v3.2.6 - the in development ARM64 version of netbox-docker
* ghcr.io/zcube/bitnami-compat/postgresql:14.4.0-debian-11-r10 - the ARM64 build of bitnami's Postgres 14
* ghcr.io/zcube/bitnami-compat/redis:7.0.3-debian-11-r14 - the ARM64 build of bitnami's Redis 7

It took some effort to get this environment up and running, partly because I think the instructions for the Kubernetes install are still based on NetBox 2.11 but, having got all the containers running, I'm now hitting the same issue/bug when I try and log in with my super user account admin/admin.

What isn't yet clear to me, and I'll try and work it out, is whether this is something to do with NetBox or something to do with the process that creates the super user in netbox-docker which I believe is somewhat different to the documented netbox install process. If there is a bug to fix, I guess I (we) need first to work out whose issues that bug should sit under!

**Reply by cpmills1975** [reply]
So, I've confirmed that the admin user that gets created by netbox-docker doesn't have a config.

I confirmed this and added it through the nbshell with the following commands:

```
>>> from django.contrib.auth.models import User

>>> User.objects.all()
<QuerySet [<User: admin>]>
>>> from users.models import UserConfig
>>> UserConfig.objects.all()
<QuerySet []>
>>> u = User.objects.get(pk=1)
>>> UserConfig(user=u, data=config.DEFAULT_USER_PREFERENCES).save()
>>> UserConfig.objects.all()
<QuerySet [<UserConfig: UserConfig object (1)>]>
```

Having done this, reloading the home page logged me in and returned what I recognise as NetBox.

Having dug around, I can see that there is a function that is supposed to be called when the user gets created. What isn't clear to me at this point is whether this should be and is being called when the super user is being created either through the method employed by netbox-docker or by the createsuperuser command.

**Reply by jeremystretch** [reply]
The UserConfig instance should be created automatically by the [`create_userconfig()`](https://github.com/netbox-community/netbox/blob/b9dd654e7ad35fe1b79feb15c92cb5276253fbbc/netbox/users/models.py#L176) handler for the `post_save` signal sent by the User model. We can verify that this functions as expected both when calling `save()` on a new User instance and when creating a user via the `create()` manager method:

```python
>>> user1 = User(username='user1')
>>> user1.save()
>>> user1.config
<UserConfig: UserConfig object (7)>
>>> 
>>> user2 = User.objects.create(username='user2')
>>> user2.config
<UserConfig: UserConfig object (8)>
```

It's likely that the Docker install is using the `bulk_create()` manager method or some other means to create the admin user, which does not send the `post_save` signal and will not trigger the receiver:

```python
>>> users = (User(username='user3'),)
>>> User.objects.bulk_create(users)
[<User: user3>]
>>> users[0].config
Traceback (most recent call last):
  File "<console>", line 1, in <module>
  File "/opt/netbox/venv/lib/python3.8/site-packages/django/db/models/fields/related_descriptors.py", line 461, in __get__
    raise self.RelatedObjectDoesNotExist(
django.contrib.auth.models.User.config.RelatedObjectDoesNotExist: User has no config.
```

As a general rule, `bulk_create()` should be avoided for this and other reasons (such as bypassing model validation) where not absolutely required for performance reasons.

**Reply by cpmills1975** [reply]
This is the code they are using (in docker-entrypoint.sh):

```
  ./manage.py shell --interface python <<END
from django.contrib.auth.models import User
from users.models import Token
if not User.objects.filter(username='${SUPERUSER_NAME}'):
    u=User.objects.create_superuser('${SUPERUSER_NAME}', '${SUPERUSER_EMAIL}', '${SUPERUSER_PASSWORD}')
    Token.objects.create(user=u, key='${SUPERUSER_API_TOKEN}')
END
```

**Reply by jeremystretch** [reply]
Is the user actually being created though, or it does it already exist? [`create_superuser()`](https://github.com/django/django/blob/d4c5d2b52c897ccc07f04482d3f42f976a79223c/django/contrib/auth/models.py#L163) should work as expected (it's just a wrapper around `save()` on the instance).

```python
>>> user4 = User.objects.create_superuser('user4')
>>> user4.config
<UserConfig: UserConfig object (9)>
```

**Reply by cpmills1975** [reply]
The user must have been created as this was a brand new install with a blank DB. The one thing that does come to mind is that the Kubernetes deployment brought up three replicas simultaneously - is there any way that may have caused some sort of race condition?

The log output shows that the relevant block of code executed, but perhaps it didn't - the super users' token didn't get created either.

Hmm... I'm now wondering whether the DB wasn't empty completely. I mentioned above that getting NetBox to run was a challenge due to the instructions being based on an old version. I hit some issues with redis authentication on the cache database. Is there any chance that could have perhaps interrupted the user creation process?

**Reply by jeremystretch** [reply]
Hard to speculate on what it could be. Even if it was a race condition, it should still be getting created in the database via some other instance.

The simplest solution might be to just add a check in NetBox's login function to ensure that a UserConfig exists for the user after authentication, and if not to automatically create one.

**Reply by jeremystretch** [reply]
I've opened #9741 to add a secondary check/creation for the UserConfig instance on login. Hopefully this should avoid the issue altogether.

**Comment by jeremystretch** [comment]
Just to reiterate from the thread above, I've opened #9741 to add an additional check for a missing UserConfig during login. Although we haven't been able to pinpoint the exact issue within Docker, this addition should avoid the issue entirely.

Thanks @cpmills1975 for all your work in tracking this down!

**Comment by Mikesco3** [comment]
I'm my case, I was trying to specify a user and password other than admin/admin in the docker compose yaml. 
But then I destroyed the container and just used admin/admin to get it going and it worked.
So once in Netbox I created my super user and deleted admin.
I speculate it is something related to the script that creates the user. Is there a part that is hard coding the creation of an admin/admin or expecting a user called admin regardless of what the docker compose configuration says?

**Comment by cpmills1975** [comment]
Thanks @jeremystretch for looking in to this so quickly. I suspect my issue was an edge case, but your belt and braces approach may help others - it was hitting this error that lead me to this discussion.

So to dig further in to the root cause of this, I deleted the NetBox deployment (leaving the database untouched), changed the superuser account name, in my case in the Kubernetes configmap, but I guess the Docker compose file would be equivalent, then recreated the deployment and I can confirm that my second admin account was properly created along with a UserConfig.

**Comment by ed-ud** [comment]
This bug just started happening to me after upgrading our NetBox instance from v3.5.2 to v3.6.4.

The problem appears to be, when I create a new user, it doesn't add a new row to the users_userconfig table for that id.  I tried creating and deleting several users before I finally figured this out.  Manually adding an "empty" config row to users_userconfig with the new id appears to work around the problem.

**Comment by ed-ud** [comment]
@jeremystretch - should I open a new issue or did you want to revive the existing issue: https://github.com/netbox-community/netbox/issues/9741 ?

**Comment by kudlaty79bc** [comment]
Hello,
Have the same issue in 3.6.4 (new installation with database transfered from 3.6.3).
Using below command directly in database works
nbdb=> INSERT INTO users_userconfig values (1,'{}',1);
where "1" is ID and USER_ID

But that is definitely a bug.

**Comment by ed-ud** [comment]
I guess we'll need to create a new issue...

**Comment by Guruleenyc** [comment]
How do I fix this on a brand new install after trying to login ?
![Screenshot from 2025-05-19 09-30-03](https://github.com/user-attachments/assets/5dacf977-e277-4b53-b615-bdb2a554609a)
